8 testing episode (#9)
SS-NES/SS-NES-lesson: episodes/testing.md

```python
@pytest.mark.parametrize(
    ("onset", "phenomenon", "expected"),
    [
        (
            "2024-12-09T11:31:14Z",
            "snow-ice",
            "Monday 9 December: chance of snow/road icing",
        ),
        (
            "2025-01-04T00:00:00Z",
            "low-temperature",
            "Saturday 4 January: chance of cold",
        ),
    ],
    ids=["special_case", "normal_case"],
)
def test_get_english_headline(onset: str, phenomenon: str, expected: str) -> None:
    """test generation of english headline"""
    assert _get_english_headline({"onset": onset, "phenomenon": phenomenon}) == expected
```

As you can see even the expected result is now an input of the test. We can use the ids parameter to give a test a name. With this name you can also run the test for only one of the ids.

# 2. Testing a unit of software without having to instantiate the all the code

> Sometimes it happens that you want to test function but in it a lot of complex objects are used that also need other objects. One way to deal with this is to add those complex objects as input to the function. You can that use this mock to prevent you having to create all those objects yourself.

Reviewer: ```suggestion
Sometimes it happens that you want to test a function but in that function a lot of complex objects are used (and those objects in turn need other objects...). One way to deal with this is to add those complex objects as input to the function. You can that use this mock to prevent you having to create all those objects yourself.
```

> In the code bellow we see the complex class being mocked and then given an implementation for when the method is called. This way we don't need to create input_one and input_two with all of their possible inputs. This type of test double tests state and behaviour.

Reviewer: ```suggestion
In the code bellow we see the complex class being mocked and then given an implementation for when the method is called. This way we don't need to create `input_one `and `input_two `with all of their possible inputs. This type of test double tests state and behaviour.
```

```python
from unittest.mock import MagicMock

class Complex:
    
    def __init__(self, input_one, input_two):
        self.input_one = input_one
        self.input_two = input_two
    
    def execute(self):
        "do complex things"
        pass
    
def function_under_test(my_complex_object_with_multiple_inputs):
    return my_complex_object_with_multiple_inputs.execute()

def test_function_under_test():
    inputs = MagicMock()
    inputs.execute = MagicMock(return_value=3)
    result = function_under_test(inputs)
    expected = 3
    assert result == expected
    assert inputs.execute.call_count == 1
```

> 

Reviewer: Again, we could point to further reading
```suggestion
For more information on mocking you can read [this quick guide](https://docs.python.org/3/library/unittest.mock.html#quick-guide).
```

Another way to not have to create the whole object graph is to use a fake. A fake a class that looks like the class it fakes (so inheritance is possible here) and has the methods that we want to test override. This type of test double tests state.

```python
class Complex:

    def __init__(self, input_one, input_two):
        self.input_one = input_one
        self.input_two = input_two

    def execute(self):
        "do complex things"
        pass

def function_under_test(my_complex_object_with_multiple_inputs):
    return my_complex_object_with_multiple_inputs.execute()

class ComplexFake(Complex):
    def __init__(self):
        pass

    def execute(self):
        return 3

def test_function_under_test():
    inputs = ComplexFake()
    result = function_under_test(inputs)
    expected = 3
    assert result == expected
```

> 

Reviewer: Is there a preference between using Fake functions and using Mock objects? I would be in favour of having one "preferred" way we recommend.

Author: When we want to show only one I would say that we use a Mock. A fake could save you a library, but most people nowadays already use methods / objects from testing libraries anyways. Shale I remove the part about the Fake?

Reviewer: Yes, let's do that, thanks!

Author: removed it

def get_konrad3d_data(url):
    response = requests.get(url)
    return response.text

def extract_latest_file(overview_page):
    soup = BeautifulSoup(overview_page, features="html.parser")
    urls = soup.find_all('a')
    latest_file = urls[-1].get('href')
    return latest_file

def get_latest_file_konrad3d():
    url = "https://opendata.dwd.de/weather/radar/konrad3d/"
    overview_page = get_konrad3d_data(url)
    latest_file = extract_latest_file(overview_page)
    return latest_file

data = '<html><head><title>Index of /weather/radar/konrad3d/</title></head><body><h1>Index of /weather/radar/konrad3d/</h1><hr><pre><a href="../">../</a><a href="KONRAD3D_20241116T093000.xml">KONRAD3D_20241116T093000.xml</a>                       16-Nov-2024 09:34                3895<a href="KONRAD3D_20241118T092500.xml">KONRAD3D_20241118T092500.xml</a>                       18-Nov-2024 09:30                3938</pre><hr></body></html>'

def test_download_latest_data_konrad3d():
    with patch("faketest.get_konrad3d_data", return_value=data):
        result = get_latest_file_konrad3d()
    expected = "KONRAD3D_20241118T092500.xml"
    assert result == expected
```

> Another way to not do these API calls is by using the [requests_mock library](https://pypi.org/project/requests-mock/) to mock requests API calls. This makes you dependent on another library and still does not show you if things work in reality. It's being used by a lot of people, but personally I prefer fewer dependencies and write integration tests for the integration with external systems.  When you mock this it can give you a false sense of security like happened with the Crowdstrike outage in their testing. If you want to use this an example can be found bellow.

Reviewer: Here it might be good to make the distinction between unit and integration test. I think mocking API calls is a good way to make sure your unit tests work, even if you have network issues, then you can have integration tests to make sure your pipeline still work when calling the real service.

The Turing Way has a section on integration testing, which we could point to.
https://book.the-turing-way.org/reproducible-research/testing/testing-integrationtest

Author: Added part of turing way
Changed the name of the test we do now to unit test

```python
import time
def function_to_test(duration=1):
    time.sleep(duration)
    return 123

def test_my_function(benchmark):
    allowed_speed = 1.000002
    result = benchmark.pedantic(function_to_test, iterations=5)
    assert benchmark.stats.stats.max < allowed_speed

    assert result == 123
```

This code can be run with the following command: `pytest -v -s the file_this_is_in.py::test_my_function`. It will run the code 5 times and none of the calls is allowed to be slower than the allowed_speed.

When you write API's you can also have performance requirements. For this another type of tool is used. One of the most used tools for this in python is locust. FOr more information about this tool look you can look at [their documentation](https://docs.locust.io/en/stable/what-is-locust.html).

> # 5. Smoke testing to see if your application is still doing its basic functionality

Reviewer: Maybe we can replace this section by a reference to https://book.the-turing-way.org/reproducible-research/testing/testing-smoketest

Is there anything domain-specific that we could add (examples of smoke tests in a specific domain)?

Author: Added example
Add reference

There are moments that you want to start an application but the application has some prerequisites it needs to have before you can say that it's good and allowed to run. For this you can use a smoketests. For example when you have an application that when a user calls it reads configurations files from a file system the check could be if the files exist at the correct location and the format is as expected. Maybe someone manually moved the files it this could break the whole system. So when the files are not there, there is smoke and thus if it's production we could get a fire.

> # 6. Runtime testing

Reviewer: Again, we can point to existing section https://book.the-turing-way.org/reproducible-research/testing/testing-runtime

I like the example of http requests taking 10+ seconds to reach their destination. Maybe we can elaborate more on that example? I think that is something which researchers might be able to relate to and understand better when / how to use runtime testing.

Author: Added part of turing way.
Added part where in research and time base situation you might want to use this.
